### ISSUE TITLE
The script `install_agda_user.bat` for Windows uses the dangerous `setx` command

### ISSUE THREAD

**Issue by ruifengx** [agda/agda#7562]
The `install_agda_user.bat` script is packed as part of the nightly build and released on GitHub. It uses the `setx` command to modify the `PATH` environment variable:

https://github.com/agda/agda/blob/c4f51a79650697d072aa4ffabae987e470d5c863/.github/install_agda_user.bat#L7

However, [the documentation for `setx`](https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/setx#remarks) points out the following (emphasis mine):
> Be aware there's a limit of 1024 characters when assigning contents to a variable using setx.
> 
> This means that **the content is cropped if you go over 1024 characters**, and that the cropped text is what's applied to the target variable. If this cropped text is applied to an existing variable, **it can result in loss of data previously held by the target variable.**

It is pretty common for developers to have a `PATH` variable longer than 1024 characters (in fact, mine is longer than 2048 characters), so I think it is reasonable to warn against using this script, or even remove it altogether, to avoid corrupting the user's `PATH` accidentally. Unfortunately, I do not know enough about cmd programming for a better alternative, but I am pretty sure that there should be much safer ways to edit `PATH` using e.g. PowerShell.

**Comment by andreasabel** [comment]
@ruifengx Would you be willing to submit a PR that fixes the problem, e.g. by shipping a powershell script instead of the .bat?

**Comment by ruifengx** [comment]
Yes. I need some time to do some research and experiments, but I guess at least we can do the same thing as `check_agda.sh`: print what needs to be modified and leave them to the user.

**Comment by ruifengx** [comment]
Sorry for leaving this here for so long. I feel I need to post a quick update here for the current situation. These things have been stuck in my head for long enough, and maybe I should at least write a "good-enough" version to replace the current batch script.

There are two API functions `[System.Environment]::{Get,Set}EnvironmentVariable` from .NET Framework. However, they also have their own gotchas:
1. Calling `GetEnvironmentVariable` fully expands the variable, so references to other variable like `%SOME_VARIABLE%` are expanded into plain text;
2. Calling `SetEnvironmentVariable` changes the registry type from `REG_EXPAND_SZ` to `REG_SZ`, so should references like `%SOME_VARIABLE%` be added to `PATH` later, it would not be automatically expanded by Windows any more;

These issues are known but there is no plan to fix (cf. dotnet/runtime#1442); even if it gets fixed, it is better not to rely on the fix due to the risk.

So the only option that I know of is to directly edit the Windows registry (which requires even more care). If we trust this StackOverflow answer [^1] and this blog post [^2], here is what we should do in general:
[^1]: https://stackoverflow.com/a/27524483
[^2]: https://web.archive.org/web/20120103121652/http://blogs.technet.com/b/heyscriptingguy/archive/2011/07/23/use-powershell-to-modify-your-environmental-path.aspx
```
$oldPath = (Get-ItemProperty -Path 'Registry::HKEY_LOCAL_MACHINE\System\CurrentControlSet\Control\Session Manager\Environment' -Name PATH).Path
$newPath = $oldPath + ’;’ + $pathToAdd
Set-ItemProperty -Path 'Registry::HKEY_LOCAL_MACHINE\System\CurrentControlSet\Control\Session Manager\Environment' -Name PATH –Value $newPath
```

However, there are still something to be careful about: the above code snippet appends `$pathToAdd` unconditionally. If the user run the script multiple times, there will be multiple copies of that path in `PATH`. I want to avoid this if possible, but that of course also makes the script more complicated. I have been working on Linux most of the time for the past months, so I did not find an opportunity to write and properly test the script. I will try to give it a shot in the upcoming holiday, and I will report back here once I have something to share.

**Comment by andreasabel** [comment]
The current content of the script is this one: https://github.com/agda/agda/blob/c0c969c661c6f2513151b6655f54e0fd99edb98d/.github/install_agda_user.bat#L1-L18

Since the `Agda_datadir` variable needs no longer be set to run agda (from 2.8.0 up), this script might become entirely obsolete, unless we want to set the PATH for the user...

**Comment by ruifengx** [comment]
That's a really good news to hear! I think the most reliable way of configuring the environment variables is to have the user do it themselves, due to all the complications of the Windows registry. In this case, I'd suggest we remove the script and close this issue.

**Comment by andreasabel** [comment]
Summoning @L-TChen as the author of these scripts to weigh in his opinion...

**Comment by L-TChen** [comment]
I think this script was originally written by @ice1000, if I remember correctly.

**Comment by ice1000** [comment]
> The current content of the script is this one: https://github.com/agda/agda/blob/c0c969c661c6f2513151b6655f54e0fd99edb98d/.github/install_agda_user.bat#L1-L18
> 
> Since the `Agda_datadir` variable needs no longer be set to run agda (from 2.8.0 up), this script might become entirely obsolete, unless we want to set the PATH for the user... 

Yes, the script was made to set Agda_datadir correctly (and do some sanity checks). And I'm glad to see we no longer need it!

**Comment by andreasabel** [comment]
I am still a bit clueless on this issue because I do not really understand how the dists generated by our deploy script are supposed to be used.  
- Who actually uses them?
- Does someone has any tests for them?  (We do not seem to have such.)

This is the part generating the dists: https://github.com/agda/agda/blob/493001846fbf63a228381e515b61f73b35d7d58f/src/github/workflows/deploy.yml#L251-L265

The structure created is
- some shell/batch scripts
- directory `bin/` with binaries `agda` and `agda-mode`
- directory `data/` which is made target of `Agda_datadir` by the scripts

1. We could get rid of `data/` because this is part of `agda` but also we need not, since one can still use the `Agda_datadir` variable now.
2. We may want to keep the `agda-mode` executable so that switching in Emacs from old agda versions to new agda versions continues to work.
3. We could get rid of the install scripts but we need not.

It would be good if the dists created by the deploy script had a maintainer that has more clue about their purpose and functioning than me.
@L-TChen would you be suitable?

If it was just me I would trash the whole thing and just ship the single `agda` executable instead of the archive with bin, data, and scripts.

**Comment by L-TChen** [comment]
> 1. We could get rid of `data/` because this is part of `agda` but also we need not, since one can still use the `Agda_datadir` variable now.
> 2. We may want to keep the `agda-mode` executable so that switching in Emacs from old agda versions to new agda versions continues to work.
> 3. We could get rid of the install scripts but we need not.
> 
> It would be good if the dists created by the deploy script had a maintainer that has more clue about their purpose and functioning than me. [@L-TChen](https://github.com/L-TChen) would you be suitable?
> 
> If it was just me I would trash the whole thing and just ship the single `agda` executable instead of the archive with bin, data, and scripts.

Indeed, we (or I?) should simplify the script for the distribution generation which was created to fix a few issues: 

1. macOS needs notarization to run; otherwise, the user needs to turn off the security warning explicitly. This is inherent to any binary compiled by a third party rather than a local host. 
2. The use of `Agda_datadir` -- the script sets up some environment variables.
3. Run `agda-mode` to set up the Emacs mode.

The second issue is gone completely, and I am skeptical that one needs a script for setting up Emacs.

> Who actually uses them?

@banacorn mentioned to me that he plans to fetch binaries directly from these releases for the VS code extension.

> Does someone has any tests for them? (We do not seem to have such.)

No. I propose to simplify the deploy script and add a test for them with a set of instructions to use them in the manual. I should be able to do it, say, in May. Does it sound reasonable?

**Comment by andreasabel** [comment]
> I propose to simplify the deploy script and add a test for them with a set of instructions to use them in the manual. I should be able to do it, say, in May. Does it sound reasonable?

That would be great, thanks!  I assign you to this issue.

Depending how quickly you get to it, this could be in 2.8.0, but if I understand correctly, this would not hold up the release of 2.8.0, am I right?

**Comment by ice1000** [comment]
I also use these dists myself. It's insanely useful for Haskell-free environments.

**Comment by andreasabel** [comment]
Ok, then I think we have a setup: @L-TChen maintains the dists and @ice1000 can be beta-tester and advisor.

**Comment by L-TChen** [comment]
Question: 

Shall we compile these dists with the flag `enable-cluster-counting`? 

The ICU library cannot be statically linked, so they were copied as as part of the distribution. The linux binary is not compiled with this flag in order to be statically linked. Perhaps it is more consistent not to have this flag.

**Comment by andreasabel** [comment]
> Perhaps it is more consistent not to have this flag.

Makes sense.
Greetings from AIM XL in Budapest!
